internal/tunnel/client/raw_session.go

```diff
@@ -268,6 +278,15 @@ func (s *rawSession) rpc(reqtype proto.ReqType, req any, resp any) error {
 }
 
 func (s *rawSession) onHeartbeat(pingTime time.Duration, timeout bool) {
+	// make sure we don't send on a closed channel.
+	// Any number of `onHeartbeat` callbacks can be in flight at a given time,
+	// but only one Close.
+	s.closedLock.RLock()
```

Comment:
This ends up with a deadlock in the `s.Close` before when it tries to take the `Lock`, right?

Manually scoping the lock by unlocking it before calling s.Close probably fixes that easily enough, and I think overall does make this simpler, good idea on modeling it this way
Still, unit tests feel very warranted here!

Reply:
Whoops, you are correct! Overlooked that `onHeartbeat` calls `Close` itself.